feat: switch to sqlite instead of json for state persistence

cargopete opened the pull request:
### Description

- [x] PersistedState struct removed in favour of passing a db pool reference (`&SqlitePool`) wherever we need to update/query the db 
- [x] updated unit tests (they were on the state file before)
- [x] updated e2e tests

hopeyen commented on subgraph-radio/src/config.rs:
```diff
@@ -306,7 +283,7 @@ pub struct RadioSetup {
         long,
         value_name = "GRAPHCAST_NETWORK",
-        default_value = "testnet",
+        default_value = "mainnet",
```
let's also update other network configs (network and registry subgraph) if we want to use `mainnet` as a default

hopeyen commented on subgraph-radio/src/entities/mod.rs:
```diff
@@ -0,0 +1,737 @@
+    pool: &SqlitePool,
+    new_attestation: NewAttestation,
+) -> Result<AttestationRecord, SqlxError> {
+    let senders = new_attestation.senders.join(",");
```
in contrast of using delimiters, what are your thoughts about using json serialization that preserves the vector format? (though the current way should be robust enough for our use case

hopeyen commented on subgraph-radio/src/entities/mod.rs:
```diff
@@ -0,0 +1,737 @@
+    comparison_result: &ComparisonResult,
+) -> Result<(), SqlxError> {
+    let local_attestation_json =
+        serde_json::to_string(&comparison_result.local_attestation.as_ref()).unwrap();
```
return Error instead of `unwrap()`, update the result error type
also the line below

hopeyen commented on subgraph-radio/migrations/20231109112858_persisted-state.up.sql:
Could you also make metadata files to [enable offline build mode](https://github.com/launchbadge/sqlx/blob/main/sqlx-cli/README.md#enable-building-in-offline-mode-with-query)?

hopeyen commented on subgraph-radio/src/entities/mod.rs:
```diff
@@ -0,0 +1,737 @@
+    let key = msg.payload.subgraph_id.clone();
+    let existing_message = get_upgrade_intent_message(pool, &key).await?;
+
+    match existing_message {
```
it might be quicker to use `UPSERT` operation with `INSERT ... ON CONFLICT` 
```
INSERT INTO upgrade_intent_messages (identifier, nonce, graph_account, ...)
VALUES (?, ?, ?, ...)
ON CONFLICT (deployment)
DO UPDATE SET
    identifier = excluded.identifier,
    nonce = excluded.nonce,
    graph_account = excluded.graph_account,
    ...;
```

then this single query should be able to handle both adding messages and replacing existing messages

hopeyen commented on subgraph-radio/src/entities/mod.rs:
```diff
@@ -0,0 +1,737 @@
+
+impl ComparisonResultRecord {
+    pub fn into_comparison_result(self) -> Result<ComparisonResult, serde_json::Error> {
+        let result_type = match self.result_type.as_str() {
```
there should be an existing `from_str` for this enum?

hopeyen commented on subgraph-radio/src/lib.rs:
```diff
@@ -222,6 +253,98 @@ impl ControlFlow {
     }
 }
 
+pub async fn create_test_db(connection_string: Option<&str>) -> SqlitePool {
```
perhaps utilize `sqlx::test` macro so you could call something like 
```
#[sqlx::test(migrations = "../migrations")]
```

cargopete replied:
tried this and it doesn't seem to be working since the migrations folder is in a different workspace member. I could try to move it to the root of the workspace if it sounds ok

cargopete replied:
Update - I had a ton of problems moving the migrations folder to root, would rather not spend too much time on this if it's ok, I don't think having a helper function for this for tests is that bad

hopeyen replied:
hmm what was the problem? could the path be defined differently to wherever the migration folder is? I think using the macro will make the tests a lot easier to maintain and utilize the true database setup

hopeyen commented on subgraph-radio/src/messages/poi.rs:
```diff
@@ -249,21 +244,33 @@ pub async fn send_poi_message(
-                    );
-                    trace!("save local attestations: {:#?}", local_attestations);
+                    // After successfully sending, save the attestation
+                    let new_attestation = NewAttestation {
```
Let's try to accommodate the usage of `NewAttestation` within `Attestation` so we don't have overlapping types like this

hopeyen commented on subgraph-radio/src/messages/poi.rs:
```diff
@@ -288,93 +295,100 @@ pub async fn send_poi_message(
-            .unwrap(),
-    );
+    // Create a NewRemotePpoiMessage from GraphcastMessage<PublicPoiMessage>
+    let new_message = NewRemotePpoiMessage {
```
do you think we can move the type conversion into db helpers? meaning we pass `msg` into `insert_remote_ppoi_message` directly and have it handle the conversions instead of converting here?

hopeyen commented on subgraph-radio/src/entities/mod.rs:
```diff
@@ -0,0 +1,737 @@
+    Ok(attestation)
+}
+
+pub async fn get_all_local_attestations_for_identifier(
```
maybe rename this to something like `get_local_attestations_by_identifier`?

hopeyen commented on subgraph-radio/src/operator/attestation.rs:
```diff
@@ -185,94 +223,55 @@ pub fn combine_senders(attestations: &[Attestation]) -> Vec<String> {
+    // Fetch all local attestations for the identifier from the database
+    let local_attestations = get_all_local_attestations_for_identifier(&db, id).await?;
+
+    let remote_blocks: HashSet<u64> = remote_ppoi_messages
```
it seems to me that we don't need to pass `remote_ppoi_messages` into this function, instead just query it from the DB with the `id` filter. Maybe look at where else are messages used and decide if it is better to pass the reference around or just make a db query?

hopeyen commented on subgraph-radio/src/operator/attestation.rs:
```diff
@@ -185,94 +223,55 @@ pub fn combine_senders(attestations: &[Attestation]) -> Vec<String> {
-    }
-}
+        })
+        .min_by_key(|&(block_number, timestamp)| (block_number, timestamp));
```
I think performance is better if we get the min first and then map `+ collection duration`

hopeyen commented on subgraph-radio/src/operator/attestation.rs:
```diff
@@ -305,6 +304,86 @@ impl ComparisonResult {
+        .into_iter()
+        .next();
+
+    let result_type = match existing_result {
```
An idea to improve this block is to make `update_comparison_result` return a bool for whether it made an update or the number of rows it updated, and use that to determine `should_notify`. `update_comparison_result` could be using `insert ... on conflict` as well

hopeyen commented on subgraph-radio/src/operator/attestation.rs:
```diff
@@ -384,124 +463,54 @@ impl Clone for ComparisonResult {
-/// The top remote attestation is found by grouping attestations together and increasing their total stake-weight every time we see a new message
-/// with the same PPOI from an Indexer (NOTE: one Indexer can only send 1 attestation per subgraph per block). The attestations are then sorted
-/// and we take the one with the highest total stake-weight.
+/// Compares a local attestation against remote ones.
```
It is good to describe the parameters, I think we should still keep some descriptions for how attestations are compared

hopeyen commented on subgraph-radio/src/entities/mod.rs:
```diff
@@ -0,0 +1,737 @@
+    Ok(notification)
+}
+
+pub async fn get_all_notifications(
```
imo we don't need to explicitly say `all` in a getter function. it should be clear from the fact that there's no filter input params

hopeyen commented on subgraph-radio/src/server/model/mod.rs:
```diff
@@ -55,16 +64,13 @@ impl QueryRoot {
-            .data_unchecked::<Arc<SubgraphRadioContext>>()
-            .local_attestations(identifier, block);
-        let filtered = attestations_to_vec(&attestations);
+        let pool = &RADIO_OPERATOR.get().unwrap().db;
```
`SubgraphRadioContext` should have db access, why remove it and add `RADIO_OPERATOR`?

hopeyen commented on subgraph-radio/src/server/model/mod.rs:
```diff
@@ -201,72 +208,66 @@ pub fn calc_ratios(
-    pub fn remote_ppoi_messages(&self) -> Vec<GraphcastMessage<PublicPoiMessage>> {
-        self.persisted_state.remote_ppoi_messages()
+    pub async fn remote_ppoi_messages(&self) -> Vec<GraphcastMessage<PublicPoiMessage>> {
+        get_remote_ppoi_messages(&RADIO_OPERATOR.get().unwrap().db)
```
here as well, `&self` should already have `db` field, right? Any particular reason why `RADIO_OPERATOR` is used instead?

hopeyen commented on subgraph-radio/src/server/model/mod.rs:
```diff
@@ -275,25 +276,64 @@ impl SubgraphRadioContext {
+    pub async fn upgrade_intent_messages(
+        pool: &SqlitePool,
+    ) -> Result<HashMap<String, GraphcastMessage<UpgradeIntentMessage>>, SqlxError> {
+        let records = sqlx::query_as!(
```
use a helper function from `entities`?

hopeyen reviewed (commented):
looks good overall!

hopeyen commented on subgraph-radio/migrations/20231204080834_persisted-state.up.sql:
```diff
@@ -0,0 +1,61 @@
+    UNIQUE (block_number, identifier, ppoi)
+);
+
+CREATE TABLE IF NOT EXISTS local_attestation_senders (
```
Could you explain the rationale of defining `local_attestation_senders` and `local_attestation_timestamps` as a separate table? I think we talked about serializing senders and timestamps within an entry of attestation?

cargopete replied:
i thought this would improve performance, to be honest i could do either, let me know what you prefer

hopeyen replied:
let's keep using a single table for `local_attestation`

hopeyen commented on subgraph-radio/src/entities/mod.rs:
```diff
@@ -0,0 +1,786 @@
+        results.push(ComparisonResult {
+            deployment: row.deployment,
+            block_number: row.block_number as u64,
+            result_type: ComparisonResultType::from_str(&row.result_type).unwrap(),
```
avoid `unwrap()`?

hopeyen commented on subgraph-radio/src/entities/mod.rs:
```diff
@@ -0,0 +1,786 @@
+
+#[derive(sqlx::FromRow)]
+#[allow(dead_code)]
+pub struct RemotePpoiMessageRecord {
```
Would you please document why this type is still defined? If it is due to `PublickPoiMessage`'s block_number: u64`, I think we can update its type to be u32

cargopete replied:
i think it is because of the signature field, since we use `PublicPoiMessage` both when we are receiving and sending a message, when receiving it's fine since we have the signature on the outer GraphcastMessage struct, but when we construct a `PublicPoiMessage` to send we do not have the signature value, since it gets added by the SDK

cargopete replied:
that is why I added signature to UpgradeIntentMessage but that's ok since those types of messages are only being received by the Radio

hopeyen replied:
hmm where are you adding local messages to `RemotePpoiMessageRecord`? To have the `signature` field, did you try to wrap the radio message type with `GraphcastMessage<T>`?

hopeyen commented on subgraph-radio/src/messages/poi.rs:
```diff
@@ -249,21 +244,33 @@ pub async fn send_poi_message(
+                        error!(
+                            err = tracing::field::debug(&e),
+                            "Failed to save local attestation"
+                        );
```
perhaps return `Err` after this so it doesn't continue with the `trace!` log and `Ok(msg_id)`?

hopeyen commented on subgraph-radio/src/entities/mod.rs:
maybe rename this folder to `database`? entities is a bit more general

hopeyen commented on subgraph-radio/src/entities/mod.rs:
```diff
@@ -0,0 +1,786 @@
+    Ok(full_attestations)
+}
+
+pub async fn get_all_local_attestations(pool: &SqlitePool) -> Result<Vec<Attestation>, SqlxError> {
```
maybe just call this `get_local_attestations`?

hopeyen commented on subgraph-radio/src/operator/operation.rs:
```diff
@@ -119,59 +124,68 @@ impl RadioOperator {
-                    local_attestations,
-                )
-                .await
+                poi_message_comparison(id.clone(), collect_duration, callbook, filtered_msg, db)
```
similar to what you did with `local_comparison_point`, `poi_message_comparison` does not need to have `filtered_msg` passed in. Perhaps utilize `db` connection within `poi_message_comparison` to do a `get_remote_ppoi_messages_by_identifier`?

hopeyen commented on subgraph-radio/src/operator/attestation.rs:
```diff
@@ -185,94 +184,59 @@ pub fn combine_senders(attestations: &[Attestation]) -> Vec<String> {
+            attestation
+                .timestamp
+                .first()
+                .map_or(i64::MAX, |&timestamp| {
```
is it possible to update timestamp type to something more sensible, like `u32`?

hopeyen commented on subgraph-radio/src/operator/attestation.rs:
```diff
@@ -185,94 +184,59 @@ pub fn combine_senders(attestations: &[Attestation]) -> Vec<String> {
-    }
-}
+    db: SqlitePool,
+) -> Result<Option<(u64, i64)>, SqlxError> {
```
The return type can be simplified to `Result<(u64, u32), SomeError>` so if it is no comparison return some Error enum

hopeyen commented on subgraph-radio/src/server/model/mod.rs:
```diff
@@ -201,72 +212,68 @@ pub fn calc_ratios(
-                }
-            } else {
-                empty_attestations
+    ) -> Vec<Attestation> {
```
What is the consideration for the changes here? Asking because I'm wondering if it is possible to reference `&self.db` instead of placing `&self` with `&SqlitePool`, particularly since there is a `&SqlitePool` field in self?

cargopete replied:
i  tried that but cannot do it because of this function
```Rust
        &self,
        ctx: &Context<'_>,
        identifier: Option<String>,
        block: Option<u64>,
    ) -> Result<Vec<Attestation>, HttpServiceError> {
        let pool = &ctx.data_unchecked::<Arc<SubgraphRadioContext>>().db;
        let attestations = SubgraphRadioContext::local_attestations(pool, identifier, block).await;

        Ok(attestations)
    }
 ```
 I tried this
 ```rust
     async fn local_attestations(
        &self,
        ctx: &Context<'_>,
        identifier: Option<String>,
        block: Option<u64>,
    ) -> Result<Vec<Attestation>, HttpServiceError> {
        let ctx = ctx.data_unchecked::<Arc<SubgraphRadioContext>>();
        let attestations = ctx.local_attestations(identifier, block).await;

        Ok(attestations)
    }
```

but it will not work because `SubgraphRadioContext` does not implement `Copy`, and we cannot clone it.

hopeyen replied:
try to modify SubgraphRadioContext's local attestations by directly accessing the pool reference
```
    pub async fn local_attestations(
        &self,
        identifier: Option<String>,
        block: Option<u64>,
    ) -> Vec<Attestation> {
        let db_records: Vec<Attestation> = match (identifier, block) {
            (Some(deployment_id), Some(block_num)) => {
                get_local_attestation(&self.db, &deployment_id, block_num)
                    .await
                    .unwrap_or_default()
                    .into_iter()
                    .collect()
            }
            (Some(deployment_id), None) => {
                get_local_attestations_by_identifier(&self.db, &deployment_id)
                    .await
                    .unwrap_or_default()
            }
            (None, None) => get_local_attestations(&self.db).await.unwrap_or_default(),
            (None, Some(_)) => Vec::new(),
        };

        db_records.into_iter().collect()
    }
```

and then use in QueryRoot resolvers
```

    async fn local_attestations(
        &self,
        ctx: &Context<'_>,
        identifier: Option<String>,
        block: Option<u64>,
    ) -> Result<Vec<Attestation>, HttpServiceError> {
        let attestations = ctx.data_unchecked::<Arc<SubgraphRadioContext>>().local_attestations(identifier, block).await.clone();

        Ok(attestations)
    }


```


similar with comparison results

```
/* Query root */

    /// Function that optionally takes in identifier and block filters.
    async fn comparison_results(
        &self,
        ctx: &Context<'_>,
        identifier: Option<String>,
        block: Option<u64>,
        result_type: Option<ComparisonResultType>,
    ) -> Result<Vec<ComparisonResult>, HttpServiceError> {
        let res = &ctx
            .data_unchecked::<Arc<SubgraphRadioContext>>()
            .comparison_results(identifier, block, result_type)
            .await;

        Ok(res.to_vec())
    }

    /// Function to grab the latest relevant comparison result of a deployment
    async fn comparison_result(
        &self,
        ctx: &Context<'_>,
        identifier: String,
    ) -> Result<Option<ComparisonResult>, HttpServiceError> {
        let res = &ctx.data_unchecked::<Arc<SubgraphRadioContext>>().comparison_result( &identifier).await;
        Ok(res.clone())
    }


/* Subgraph radio context */


    pub async fn comparison_result(
        &self,
        deployment_hash: &str,
    ) -> Option<ComparisonResult> {
        let records = get_comparison_results_by_deployment(&self.db, deployment_hash)
            .await
            .ok()?;

        let record = records.into_iter().next()?;
        Some(record)
    }

    pub async fn comparison_results(
        &self,
        identifier: Option<String>,
        block: Option<u64>,
        result_type: Option<ComparisonResultType>,
    ) -> Vec<ComparisonResult> {
        if let Some(block) = block {
            let locals =
                self.local_attestations(identifier.clone(), Some(block))
                    .await;

```


let me know if you run into problems

hopeyen commented on subgraph-radio/src/server/model/mod.rs:
```diff
@@ -201,72 +212,68 @@ pub fn calc_ratios(
+        &self,
+        pool: &SqlitePool,
+    ) -> Vec<GraphcastMessage<PublicPoiMessage>> {
+        get_remote_ppoi_messages(pool).await.unwrap_or_default()
```
are you able to reference `&self.db` instead of using an extra input?

cargopete replied:
no, for the same reason as the above comment, because of this
```rust
    async fn public_poi_messages(
        &self,
        ctx: &Context<'_>,
        identifier: Option<String>,
        block: Option<u64>,
    ) -> Result<Vec<GraphcastMessage<PublicPoiMessage>>, HttpServiceError> {
        let pool = &ctx.data_unchecked::<Arc<SubgraphRadioContext>>().db;

        let msgs = ctx
            .data_unchecked::<Arc<SubgraphRadioContext>>()
            .remote_ppoi_messages_filtered(&identifier, &block, pool)
            .await;
        Ok(msgs)
    }
```
on `QueryRoot`

hopeyen replied:
You can try this
```
impl QueryRoot {
    async fn public_poi_messages(
        &self,
        ctx: &Context<'_>,
        identifier: Option<String>,
        block: Option<u64>,
    ) -> Result<Vec<GraphcastMessage<PublicPoiMessage>>, HttpServiceError> {
        let msgs = ctx
            .data_unchecked::<Arc<SubgraphRadioContext>>()
            .remote_ppoi_messages_filtered(&identifier, &block)
            .await;
        Ok(msgs)
    }

/// SubgraphRadioContext

    pub async fn remote_ppoi_messages(
        &self,
    ) -> Vec<GraphcastMessage<PublicPoiMessage>> {
        get_remote_ppoi_messages(&self.db).await.unwrap_or_default()
    }

    pub async fn remote_ppoi_messages_filtered(
        &self,
        identifier: &Option<String>,
        block: &Option<u64>,
    ) -> Vec<GraphcastMessage<PublicPoiMessage>> {
        let msgs = self.remote_ppoi_messages().await;
        let filtered = msgs
            .iter()
            .filter(|&message| filter_remote_ppoi_messages(message, identifier, block))
            .cloned()
            .collect::<Vec<_>>();
        filtered
    }

```


please clean up the other functions as well

hopeyen commented on subgraph-radio/src/entities/mod.rs:
```diff
@@ -0,0 +1,786 @@
+    Ok(message)
+}
+
+pub async fn get_all_upgrade_intent_messages(
```
similar to the previous suggestions, let's remove `all` in the function names when there's no filters for input. 

(tbh, I don't even like including words like `get` bc that seems like a generic word too, but this opinion is not conventional

hopeyen commented on subgraph-radio/src/server/model/mod.rs:
```diff
@@ -275,73 +282,115 @@ impl SubgraphRadioContext {
         &self,
         subgraph_id: &Option<String>,
     ) -> Vec<GraphcastMessage<UpgradeIntentMessage>> {
+        let msgs = SubgraphRadioContext::upgrade_intent_messages(&RADIO_OPERATOR.get().unwrap().db)
```
Similar here, I think &self should let you access db, and `RADIO_OPERATOR` doesn't need to be  in this file. Please check for other functions as well

hopeyen reviewed (commented):
Probably need to update github workflows to compile with `ENV SQLX_OFFLINE=true` which forces SQLx to use the offline mode to statically check the database queries against files in the `.sqlx` directory.

hopeyen commented on .github/workflows/gen-binaries.yml:
```diff
@@ -36,26 +39,3 @@ jobs:
           asset_path: ./target/release/subgraph-radio
           asset_name: subgraph-radio-${{ github.event.release.tag_name }}-ubuntu
           asset_content_type: binary/octet-stream
-  build-macos:
```
is this not useful?

hopeyen commented on subgraph-radio/benches/attestations.rs:
```diff
@@ -91,71 +70,93 @@ mod attestation {
-        let local: HashMap<String, HashMap<u64, Attestation>> = black_box(local_attestations);
+    async fn comparison_point_bench(c: &mut Criterion) {
+        let runtime = tokio::runtime::Runtime::new().unwrap();
+        let pool = SqlitePool::connect("sqlite::memory:")
```
use `black_box()` for setup functions as much as you can to benchmark accurately. Also not sure why there's connect and migrate within the bench fn? since the bench is for comparison_point, everything else should be the setup. It is better to split other functions (`insert_...`) in separate benchmarks

hopeyen commented on subgraph-radio/src/messages/poi.rs:
```diff
@@ -249,21 +243,49 @@ pub async fn send_poi_message(
-                        message_block,
-                    );
-                    trace!("save local attestations: {:#?}", local_attestations);
+                    let timestamp_result: Result<u32, _> = Utc::now().timestamp().try_into();
```
iirc `nonce` is the true timestamp of the message, so try to use that instead of getting timestamp after sending the message

hopeyen commented on subgraph-radio/src/operator/attestation.rs:
```diff
@@ -17,36 +26,44 @@ use graphcast_sdk::{
         Attestation {
             ppoi,
-            stake_weight: stake_weight as i64,
+            stake_weight: stake_weight as u64,
```
it makes sense to me for stake_weight to stay as a `f32` to preserve the float point, could you try to store the column with sqlite type `REAL`?

cargopete replied:
I don't think it's worth to have `Attestation` have an f32 field, because of f32 conflicts with the `Eq` and `Hash` traits, we would need to implement the traits manually and convert the field to a fixed-point integer anyway

cargopete replied:
just noting down we agreed on having stake_weight and block_number as u32s so that sqlite converts them automatically with its INTEGER type

cargopete replied:
finally noting (after last convo) that all types will be u64, because of future proofing, and will be converted to/from string when inserting/querying to db

hopeyen reviewed (commented):
much clean! 

also pls link the testing issue we discussed. I also still think cargo tests should use the `sqlx::test` macro, cleaner testing environment

hopeyen commented on subgraph-radio/src/messages/poi.rs:
```diff
@@ -249,21 +243,49 @@ pub async fn send_poi_message(
-                        message_block,
-                    );
-                    trace!("save local attestations: {:#?}", local_attestations);
+                    let timestamp_result: Result<u32, _> = nonce.try_into();
```
would be good to refactor the processing here back into a similar notion of  `save_local_attestation(...)`

cargopete commented:
> much clean!
> 
> also pls link the testing issue we discussed. I also still think cargo tests should use the `sqlx::test` macro, cleaner testing environment

thanks!
I think I addressed all the comments now, I could not get sqlx::test to work even with just the cargo tests, again with the 
```
control attributes are not allowed unless the `migrate` feature is enabled and automatic test DB management is used; see docs
```

Lmk if it's fine to keep it as is. Also for the е2е tests testing Issue - I am not sure how to frame the improvements we want, I think we can leave it as a separate effort from this Issue, since it's not related to the db?

hopeyen commented:
> I could not get sqlx::test to work even with just the cargo tests, again with the
> 
> ```
> control attributes are not allowed unless the `migrate` feature is enabled and automatic test DB management is used; see docs
> ```


I read through the docs carefully and got cargo tests to work
```
    #[sqlx::test(migrations = "./migrations")]
    async fn test_db_basic(pool: SqlitePool) {
        // let pool = create_test_db(None).await;
    ...
    }
```

If you are curious, look more carefully in the docs [here](https://docs.rs/sqlx/latest/sqlx/attr.test.html#automatic-test-database-management-requires-migrate-feature) talking about signatures


I still think it would be better to create the e2e test re-arch issue and link it here, because the changes in the e2e tests right now is not the ideal which leads to the motivation for that spec'ing work. feel free to leave it a bit more general and fill in later. Some main points are to talk about why there are multiple test binaries and a separate db testing management, what was the limitations of having them within the radio crate, and some functions to be moved into the SDK


Other changes look okay

cargopete committed:
feat: switch to sqlite instead of json for state persistence

hopeyen reviewed (approved):
🤞!

cargopete commented:
@hopeyen I will link the testing Issue later today
